src/ML/src/MANNTrajectoryGenerator.cpp

```diff
@@ -0,0 +1,358 @@
+
+    bool resetContactList(const Contacts::EstimatedContact& estimatedContact,
+                          const std::string& contactName);
+};
+
+bool MANNTrajectoryGenerator::Impl::resetContactList(
+    const Contacts::EstimatedContact& estimatedContact, const std::string& contactName)
+{
+    // if the contact is not active we do not have to add in the contact list
+    if (!estimatedContact.isActive)
+    {
+        return true;
+    }
+    Contacts::PlannedContact temp;
+    temp.activationTime = estimatedContact.switchTime * slowDownFactor;
+    temp.deactivationTime = std::chrono::nanoseconds::max();
+    temp.index = estimatedContact.index;
+    temp.name = estimatedContact.name;
+
+    const Eigen::Vector3d ypr
+        = estimatedContact.pose.quat().toRotationMatrix().eulerAngles(2, 1, 0);
+    const Eigen::Vector3d position = estimatedContact.pose.translation();
+    temp.pose.translation({position[0], position[1], 0.0});
+    temp.pose.quat(Eigen::AngleAxis(ypr(0), Eigen::Vector3d::UnitZ()));
+
+    temp.type = Contacts::ContactType::FULL;
+    return this->contactListMap[contactName].addContact(temp);
+}
+
+bool MANNTrajectoryGenerator::Impl::updateContactList(
+    const std::string& footName,
+    const std::chrono::nanoseconds& currentTime,
+    const Contacts::EstimatedContact& estimatedContact)
+{
+
+    constexpr auto logPrefix = "[MANNTrajectoryGenerator::Impl::updateContactList]";
+
+    auto contactListIt = contactListMap.find(footName);
+    if (contactListIt == contactListMap.end())
+    {
+        log()->error("{} Unable to find the contact named {} in the dictionary.",
+                     logPrefix,
+                     footName);
+        return false;
+    }
+    Contacts::ContactList& contactList = contactListIt->second;
+
+    // if the contact list map is empty then this is the first contact that we have to add to the
+    // list read it as
+    // 1. if contact is active and the list is empty means that the foot just touched the ground
+    // 2. if contact is active and the last contact is not active means that the foot just touched
+    // the ground
+    if (estimatedContact.isActive
+        && (contactList.size() == 0
+            || !contactList.lastContact()->isContactActive(currentTime * this->slowDownFactor)))
+    {
+        Contacts::PlannedContact temp;
+        temp.activationTime = estimatedContact.switchTime * this->slowDownFactor;
+        temp.deactivationTime = std::chrono::nanoseconds::max();
+        temp.index = estimatedContact.index;
+        temp.name = estimatedContact.name;
+
+        // force the foot to be attached to a flat terrain
+        const Eigen::Vector3d ypr
+            = estimatedContact.pose.quat().toRotationMatrix().eulerAngles(2, 1, 0);
+        temp.pose.translation(
+            {estimatedContact.pose.translation()[0], estimatedContact.pose.translation()[1], 0.0});
+        temp.pose.quat(Eigen::AngleAxis(ypr(0), Eigen::Vector3d::UnitZ()));
```

Comment:
Same as above. Also, I would suggest defining a separate method to do this computation

Reply:
As before